controllers/kubeadmconfig_controller.go

```diff
@@ -148,32 +217,51 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 		config.Status.BootstrapData = cloudInitData
 		config.Status.Ready = true
 
-		if err := r.Update(ctx, &config); err != nil {
-			log.Error(err, "failed to update config")
+		if err := r.patchConfig(ctx, &config, patchConfig); err != nil {
+			log.Error(err, "failed to patch config")
 			return ctrl.Result{}, err
 		}
+		log.Info("KubeadmConfig updated with BootstrapData for kubeadm init; Status.Ready=true")
+
 		return ctrl.Result{}, nil
-	}
 
-	// Requeue until the machine is ready
-	if !cluster.Status.InfrastructureReady {
-		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
 	}
 
 	// Every other case it's a join scenario
+	// Nb. in this case ClusterConfiguration and JoinConfiguration should not be defined by users, but in case of misconfigurations, CAPBK simply ignore them
+
+	if config.Spec.JoinConfiguration == nil {
+		return ctrl.Result{}, errors.New("Control plane already exists for the cluster, only KubeadmConfig objects with JoinConfiguration are allowed")
+	}
+
+	if len(cluster.Status.APIEndpoints) == 0 {
+		return ctrl.Result{}, errors.New("Control plane already exists for the cluster but cluster.Status.APIEndpoints are not set")
+	}
+
+	// Injects the controlplane endpoint address
+	// NB. this assumes nodes using token discovery
+	if config.Spec.JoinConfiguration.Discovery.BootstrapToken == nil {
+		config.Spec.JoinConfiguration.Discovery.BootstrapToken = &kubeadmv1beta1.BootstrapTokenDiscovery{}
+	}
+
+	config.Spec.JoinConfiguration.Discovery.BootstrapToken.APIServerEndpoint = fmt.Sprintf("https://%s:%d", cluster.Status.APIEndpoints[0].Host, cluster.Status.APIEndpoints[0].Port)
```

Comment:
Is the bootstrap token being generated elsewhere? Or is that still an outstanding todo?

Reply:
I assume this should be generated elsewhere because I don't see how CABPK can generate a bootstrap token and ensure it gets passed to all the machine in the cluster, but I can be totally wrong.... @ncdc @chuckha suggestions?

Reply:
In CAPA v1alpha1, we generate a bootstrap token secret, store it in the apiserver, and then plumb it through to the cloud-init data for each machine. We can do the same for CABPK.

Reply:
It's not implemented yet, but the idea is exactly as what @ncdc stated

Reply:
As an aside, I'm wondering if when we implement it here we should make sure that we shorten the lifetime of the bootstrap token.